File: RiceCooker.php
<?php

class RiceCooker
{
    private $ingredients = [];
    private $temperature;
    private $duration;

    public function __construct()
    {
        $this->resetCooker();
    }

    public function start()
    {
        $this->displayMainMenu();
    }

    private function resetCooker()
    {
        $this->ingredients = [];
        $this->temperature = null;
        $this->duration = null;
    }

    private function displayMainMenu()
    {
        echo "Main Menu:\n";
        echo "1>->-> Cook Rice\n";
        echo "2>->-> Boil Water\n";
        echo "3>->-> Make Soup\n";
        echo "0>->-> Exit\n";

        $choice = $this->getNumericInput("Choose your action");

        switch ($choice) {
            case 1:
            case 2:
            case 3:
                $this->prepareRecipe($choice);
                break;
            case 0:
                echo "Exiting...\n";
                exit;
            default:
                echo "Invalid choice. Please try again.\n";
                $this->displayMainMenu();
        }
    }

    private function prepareRecipe($recipeType)
    {
        $this->resetCooker();

        echo "Preparing recipe...\n";

        $this->addIngredients();
        $this->displayRecipeResult($recipeType);

        $this->resetCooker();
        $this->displayMainMenu();
    }

    private function addIngredients()
    {
        echo "Adding ingredients:\n";

        do {
            $ingredient = $this->getStringInput(">->-> Ingredient (e.g., water): ");
            $quantity = $this->getNumericInput(">->-> Quantity in grams: ");

            $this->ingredients[] = [
                'ingredient' => $ingredient,
                'quantity' => $quantity,
            ];

            $choice = $this->getNumericInput("1>->-> Add another ingredient\n2>->-> Finish adding ingredients\n3>->-> Cancel");

            if ($choice == 2) {
                break;
            } elseif ($choice == 3) {
                echo "Cancelling...\n";
                $this->resetCooker();
                $this->displayMainMenu();
            }
        } while (true);
    }

    private function displayRecipeResult($recipeType)
    {
        echo "Recipe Result:\n";
        echo "The dish is ready!\n";
        echo "Ingredients used:\n";
        foreach ($this->ingredients as $ingredient) {
            echo "{$ingredient['quantity']} grams of {$ingredient['ingredient']}\n";
        }
        echo "Temperature: Automatic\n";
        echo "Duration: Automatic\n";
    }

    private function getNumericInput($prompt)
    {
        do {
            $input = $this->getStringInput($prompt);
            try {
                $numericValue = filter_var($input, FILTER_VALIDATE_FLOAT);
                if ($numericValue !== false && $numericValue >= 0) {
                    return $numericValue;
                } else {
                    throw new Exception("Invalid input. Please enter a non-negative number.");
                }
            } catch (Exception $e) {
                echo "Error: " . $e->getMessage() . "\n";
            }
        } while (true);
    }

    private function getStringInput($prompt)
    {
        do {
            echo "$prompt: ";
            $input = trim(fgets(STDIN));

            if ($input === false) {
                echo "Invalid input. Please try again.\n";
            } else {
                return $input;
            }
        } while (true);
    }
}

$riceCooker = new RiceCooker();
$riceCooker->start();

?>

File: ricecooker.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

// Cooker represents the state of the rice cooker.
type Cooker struct {
	pluggedIn       bool
	recipeChosen    bool
	ingredients     map[string]int
	temperature     int
	cookingDuration int
}

// NewCooker initializes a new Cooker instance.
func NewCooker() *Cooker {
	return &Cooker{
		pluggedIn:       false,
		recipeChosen:    false,
		ingredients:     make(map[string]int),
		temperature:     0,
		cookingDuration: 0,
	}
}

// ValidateInput validates if the input can be converted to a non-negative integer.
func ValidateInput(input string) (int, error) {
	value, err := strconv.Atoi(input)
	if err != nil || value < 0 {
		return 0, fmt.Errorf("invalid input, please enter a non-negative number")
	}
	return value, nil
}

// GetUserInput prompts the user for input and validates it.
func GetUserInput(prompt string) (int, error) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Print(prompt)
	scanner.Scan()
	return ValidateInput(scanner.Text())
}

// ShowMainMenu displays the main menu options.
func ShowMainMenu() {
	fmt.Println("Choose your action:")
	fmt.Println("1>->-> Cook rice")
	fmt.Println("2>->-> Boil water")
	fmt.Println("3>->-> Make soup")
}

// AddIngredient prompts the user to add an ingredient to the cooker.
func (c *Cooker) AddIngredient() {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print(">->-> Ingredient (e.g., water): ")
		scanner.Scan()
		ingredient := scanner.Text()

		quantity, err := GetUserInput(">->-> Quantity in grams: ")
		if err != nil {
			fmt.Println(err)
			continue
		}

		c.ingredients[ingredient] = quantity

		fmt.Println("1>->-> Add another ingredient")
		fmt.Println("2>->-> Close the rice cooker and start cooking")
		fmt.Println("3>->-> Cancel")

		choice, err := GetUserInput("Choose an option: ")
		if err != nil {
			fmt.Println(err)
			continue
		}

		if choice == 2 {
			break
		} else if choice == 3 {
			c.CancelOperation()
			return
		}
	}
}

// StartCooking simulates the cooking process and displays the results.
func (c *Cooker) StartCooking() {
	fmt.Println("Cooking started:")
	fmt.Println("The dish is ready!")
	fmt.Println("Ingredients used:", c.ingredients)
	fmt.Println("Temperature:", c.temperature, "°C")
	fmt.Println("Duration:", c.cookingDuration, "minutes")
	c.CancelOperation()
}

// CancelOperation resets the cooker state to its initial values.
func (c *Cooker) CancelOperation() {
	c.recipeChosen = false
	c.ingredients = make(map[string]int)
	c.temperature = 0
	c.cookingDuration = 0
}

func main() {
	cooker := NewCooker()

	for {
		ShowMainMenu()

		choice, err := GetUserInput("Choose an option: ")
		if err != nil {
			fmt.Println(err)
			continue
		}

		switch choice {
		case 1, 2, 3:
			cooker.recipeChosen = true
			cooker.temperature = 100 // Assuming a default temperature of 100°C
			cooker.cookingDuration = 30 // Assuming a default cooking duration of 30 minutes
			cooker.AddIngredient()
			cooker.StartCooking()
		default:
			fmt.Println("Invalid choice. Please try again.")
		}
	}
}
